src/middleware/externalRequests.js

```diff
@@ -63,8 +77,253 @@ const prepareRequestConfig = requestDetails => {
+}
+
+const sendMappedObject = (ctx, axiosConfig, request, body) => {
+  const reqTimestamp = new Date()
+  let response, error, responseTimestamp
+
+  return axios(axiosConfig)
+    .then(resp => {
+      response = resp
+      response.body = resp.data
+      responseTimestamp = new Date()
+
+      if (request.primary) {
+        setKoaResponseBodyFromPrimary(ctx, request, response.body)
+
+        ctx.status = response.status
+      } else {
+        setKoaResponseBodyFromPrimary(ctx, request, response.body)
+      }
+    })
+    .catch(err => {
+      responseTimestamp = new Date()
+
+      const result = handleRequestError(ctx, request, err)
+      response = result.response
+      error = result.error
+    })
+    .finally(() => {
+      if (ctx.request.header && ctx.request.header['X-OpenHIM-TransactionID']) {
+        const orch = createOrchestration(
+          request,
+          body,
+          response,
+          reqTimestamp,
+          responseTimestamp,
+          request.id,
+          error
+        )
+
+        ctx.orchestrations.push(orch)
+      }
+    })
+}
+
+const constructOpenhimResponse = (ctx, responseTimestamp) => {
+  const response = ctx.response
+  const orchestrations = ctx.orchestrations
+  const statusText = ctx.statusText
+  const respObject = {}
+
+  if (response) {
+    if (response.headers) {
+      respObject.headers = response.headers
+    }
+    if (response.status) {
+      respObject.status = response.status
+    }
+    if (response.body) {
+      respObject.body = response.body
+    }
+    if (response.timestamp) {
+      respObject.timestamp = response.timestamp
+    } else if (responseTimestamp) {
+      respObject.timestamp = responseTimestamp
+    }
+  }
+
+  if (readError) {
+    mediatorConfigJson = {
+      urn: 'undefined'
+    }
+  }
+
+  const body = {
+    'x-mediator-urn': mediatorConfigJson.urn,
+    status: statusText,
+    response: respObject,
+    orchestrations: orchestrations
+  }
+
+  ctx.body = JSON.stringify(body)
+}
+
+const setStatusText = ctx => {
+  if (ctx.primaryReqFailError) {
+    ctx.statusText = 'Failed'
+  } else if (!ctx.primaryReqFailError && ctx.secondaryFailError) {
+    ctx.statusText = 'Completed with error(s)'
+  } else if (
+    !ctx.primaryReqFailError &&
+    !ctx.secondaryFailError &&
+    (ctx.primaryCompleted || ctx.secondaryCompleted)
+  ) {
+    ctx.statusText = 'Completed'
+  } else {
+    ctx.statusText = 'Successful'
+  }
 }
 
-if (process.env.NODE_ENV == 'test') {
+if (process.env.NODE_ENV === 'test') {
+  exports.orchestrateMappingResult = orchestrateMappingResult
+  exports.setStatusText = setStatusText
+  exports.constructOpenhimResponse = constructOpenhimResponse
   exports.prepareRequestConfig = prepareRequestConfig
+  exports.setKoaResponseBody = setKoaResponseBody
+  exports.setKoaResponseBodyFromPrimary = setKoaResponseBodyFromPrimary
+  exports.createAxiosConfig = createAxiosConfig
+  exports.handleRequestError = handleRequestError
```

Comment:
@bradsawadye Could you please try using the `rewire` module for the unit tests instead of exporting internal variables. I know this is how its been done in the past but rewire is neater. Exporting all these internal variables seems dodgy - testing every single function individually can easily lead to a **fragile** test suite which we also don't want.
@BMartinos your thoughts?

Reply:
@MattyJ007 yeah sounds good to me. We can now rather make use of the rewire module for stubbing of functions instead of having them exported just for the tests.

We also dont need to test every function individually. This can be determined on a per case basis and the function complexity. Lets just test the main functionality for now and we can always add additional tests to verify the edge cases are met

Reply:
done in https://github.com/jembi/openhim-mediator-mapping/pull/36/commits/3d8d16761713de7117573cb2c180c1361e2afc83